spatial/r2/box.go

```diff
@@ -0,0 +1,101 @@
+// Copyright ©2022 The Gonum Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style
+// license that can be found in the LICENSE file.
+
+package r2
+
+import "math"
+
+// Box is a 2D bounding box. Well formed
+// Boxes have Min components smaller than
+// Max components.
+type Box struct {
+	Min, Max Vec
+}
+
+// NewBox is shorthand for Box{Min:Vec{x0,y0}, Max:Vec{x1,y1}}.
+// The sides are swapped so that the resulting Box is well formed.
+func NewBox(x0, y0, x1, y1 float64) Box {
+	return Box{
+		Min: Vec{X: math.Min(x0, x1), Y: math.Min(y0, y1)},
+		Max: Vec{X: math.Max(x0, x1), Y: math.Max(y0, y1)},
+	}
+}
+
+// Size returns the size of the Box.
+func (a Box) Size() Vec {
+	return Sub(a.Max, a.Min)
+}
+
+// Center returns the center of the Box.
+func (a Box) Center() Vec {
+	return Add(a.Min, Scale(0.5, a.Size()))
+}
+
+// IsEmpty returns true if a Box's volume is zero
+// or if a Min component is greater than its Max component.
+func (a Box) Empty() bool {
+	return a.Min.X >= a.Max.X || a.Min.Y >= a.Max.Y
+}
+
+// Vertices returns a slice of the 4 vertices
+// corresponding to each of the Box's corners.
+//
+// The order of the vertices is CCW in the XY plane
+// starting at the box minimum.
+func (a Box) Vertices() []Vec {
+	return []Vec{
+		0: a.Min,
+		1: {a.Max.X, a.Min.Y},
+		2: a.Max,
+		3: {a.Min.X, a.Max.Y},
+	}
+}
+
+// Union returns a box enclosing both the receiver and argument Boxes.
+func (a Box) Union(b Box) Box {
+	if a.Empty() {
+		return b
+	}
+	if b.Empty() {
+		return a
+	}
+	return Box{
+		Min: minElem(a.Min, b.Min),
+		Max: maxElem(a.Max, b.Max),
+	}
+}
+
+// Add adds v to the bounding box components.
+// It is the equivalent of translating the Box by v.
+func (a Box) Add(v Vec) Box {
+	return Box{Add(a.Min, v), Add(a.Max, v)}
+}
+
+// Scale returns a new Box scaled by a size vector around its center.
+// The scaling is done element wise, which is to say
+// the Box's X size is scaled by v.X.
+// Negative components of v are interpreted as zero.
+func (a Box) Scale(v Vec) Box {
+	v = maxElem(v, Vec{})
```

Comment:
Why do this? Is there a reason to disallow scaling by negative values?

Reply:
Scale creates a new box at `b`'s center with dimensions multiplied by `scale`. Negative components of `scale` yield negative dimensions, which would create a box with `Max` < `Min`. Do we want to allow creating malformed boxes? Probably not. We can either
* Interpret this input as spurious or incorrect and set the dimension to zero which would create a well formed empty box.
* Interpret that the user implies they want the absolute value of the dimension and set the dimension to the result of `math.Abs(component)`.

To me the latter seems like the most user friendly way to go about it. It also feels like bad API design since gonum would not help users detect "bad" dimension input where it exists.

Reply:
Yeah. OK.

Reply:
Sorry I incorrectly said the first felt like bad API design when I actually meant the second. Are we still on the same page?